docs/design/mono/components.md

```diff
@@ -0,0 +1,280 @@
+time, instead of requiring custom builds of the runtime.
+
+For example, each of the following experiences requires different runtime
+capabilities:
+
+- Developer inner loop on on a mobile or WebAssembly workload: The runtime
+  should include support for the interpreter, hot reload, and the diagnostic
+  server.
+- Release build iPhone app for the app store: The runtime should not include the
+  interpreter, hot reload, or the diagnostic server.
+- Release build iPhone app with `System.Reflection.Emit` support: The runtime
+  should include the interpreter, but not hot reload or the diagnostic server.
+- Line of business Android app company-wide internal beta: The runtime should
+  not include interpreter support or hot reload, but should include the
+  diagnostic server.
+
+## Building Components
+
+For each workload we choose one of two strategies for building the runtime and
+the components: we either build the components as shared libraries that are
+loaded by the runtime at execution time; or we build the components as static
+libraries that are statically linked together with the runtime (static
+library), and the application host.  We assume that workloads that would
+utilize static linking already do native linking of the application host and
+the runtime as part of the app build.
+
+The choice of which strategy to pursue depends on platform capabilities.  For
+example there is no dynamic linking on WebAssembly at this time, so static
+linking is the only option.  On iOS dynamic linking is supported for debug
+scenarios, but is disallowed in apps that want to be published to the Apple App
+Store. Thus for ios release builds we must use static linking.
+
+We can summarize the different options:
+
+| Scenario(s)                                             | Dynamic loading allowed | Component build strategy                                                                              | Disabled components                                                                     |
+| ------------------------------------------------------- | ----------------------- | ----------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------- |
+| Console, Android,<br>ios simulator,<br>ios device debug | yes                     | component stubs in runtime; component in shared libraries next to runtime; dlopen to load.            | Just leave out the component shared library from the app bundle.                        |
+| webassembly,<br>ios device release                      | no                      | component stubs in runtime; components in static libraries; embedding API calls to register non-stubs | Don’t link the component static libs.<br>Leave out the embedding API registration call. |
+
+
+
+## High level overview
+
+Each component is defined in `src/mono/mono/components`.
+
```

Comment:
Is it really necessary to add a subdirectory for this ?

Reply:
It's not necessary, but I thought it made it clearer that components are meant to be optional